### ISSUE TITLE
Question about the descriptor branch

### ISSUE THREAD

**Issue by Zhuyawen7** [cvg/SOLD2#41]
Hello, thank you for sharing!
The backbone encoding is processed by two consecutive convolutions of kernels 3 × 3 and 1 × 1, and output channels 256 and 128, to produce a h/4 ×w/4 × 128 feature descriptor map.
Coud I modify the output channels "256 and 128" to "256 and 256"? and then i will get a h/4 ×w/4 × 256 feature descriptor map. Will this improve the performance of the network, and do I need to modify other things?

**Comment by rpautrat** [comment]
Hi, yes you can do this change, and I don't think you would need to change anything else, except adapting the output branches that are using the backbone features.

But I don't think that it would increase the performance, or only in a minor way if it does. It has for example been shown for point descriptors that a dimension of 128 is enough to store all the information about local features. The only reason why it would improve things is that the backbone features are shared for detection and description, so a larger size might help.

If you try it, I would be curious to see the results.

**Comment by rpautrat** [comment]
Hi, yes everything is normal here. We use a log in the loss, so it can be negative.
The junction precision is indeed smaller, because it is a very low-level metric at the pixel level. It is thus quite hard to get good numbers there. But the Junction NMS precision will give you a better idea of the real precision of the junctions.

**Comment by Zhuyawen7** [comment]
> Hi, yes you can do this change, and I don't think you would need to change anything else, except adapting the output branches that are using the backbone features.
> 
> But I don't think that it would increase the performance, or only in a minor way if it does. It has for example been shown for point descriptors that a dimension of 128 is enough to store all the information about local features. The only reason why it would improve things is that the backbone features are shared for detection and description, so a larger size might help.
> 
> If you try it, I would be curious to see the results.

Hi,
Now my step 1 is trained for 35 epoch, but the precision has barely changed, the junction precision is still around 0.16, is this the end result and do I need to start step 2? 
Looking forward to your reply

**Comment by Zhuyawen7** [comment]
> Hi, yes everything is normal here. We use a log in the loss, so it can be negative. The junction precision is indeed smaller, because it is a very low-level metric at the pixel level. It is thus quite hard to get good numbers there. But the Junction NMS precision will give you a better idea of the real precision of the junctions.

Thank you for your reply.
Now my step 1 is trained for 35 epoch, but the precision has barely changed, the junction precision is still around 0.16 and the junction NMS is around 0.45.I want to know what value it should reach after I can start the step 2.

**Comment by rpautrat** [comment]
Yes, these are good values already and you can stop there and proceed to step 2. In general, you can just stop the training once all the metrics have converged.

**Comment by Zhuyawen7** [comment]
Thank you! I will proceed to step 2.

**Comment by monarchdemon** [comment]
> 您好，谢谢分享！ 主干编码由内核 3×3 和 1×1 的两个连续卷积以及输出通道 256 和 128 处理，以生成 ah/4×w/4×128 特征描述符映射。 我可以将输出通道“256 和 128”修改为“256 和 256”吗？然后我会得到 ah/4 × w/4 × 256 特征描述符图。这会提高网络的性能，我是否需要修改其他东西？

您好，方便得到您的操作步骤吗？可联系：benhe_cai@163.com

**Comment by Zhuyawen7** [comment]
> > 您好，谢谢分享！ 主干编码由内核 3×3 和 1×1 的两个连续卷积以及输出通道 256 和 128 处理，以生成 ah/4×w/4×128 特征描述符映射。 我可以将输出通道“256 和 128”修改为“256 和 256”吗？然后我会得到 ah/4 × w/4 × 256 特征描述符图。这会提高网络的性能，我是否需要修改其他东西？
> 
> 您好，方便得到您的操作步骤吗？可联系：[benhe_cai@163.com](mailto:benhe_cai@163.com)

I just changed 128 to 256

**Comment by Zhuyawen7** [comment]
> 是的，这些值已经很好了，您可以停止并继续进行第 2 步。通常，您可以在所有指标收敛后停止训练。



> Yes, these are good values already and you can stop there and proceed to step 2. In general, you can just stop the training once all the metrics have converged.

Hello,
I am now on step 4. I think I have a problem. The "heatmap" precision always 0, and the "junction" precision has also been very lower. Sometimes all the precision in one iteration is zero. Is this normal please?
![image](https://user-images.githubusercontent.com/100696862/156997297-17fbd048-b665-4b0a-bc94-84a9eb4cf01b.png)

**Comment by Zhuyawen7** [comment]
> Yes, these are good values already and you can stop there and proceed to step 2. In general, you can just stop the training once all the metrics have converged.

Hello,
I am now on step 4. I think I have a problem. The "heatmap" precision always 0, and the "junction" precision has also been very lower. Sometimes all the precision in one iteration is zero. Is this normal please?
![image](https://user-images.githubusercontent.com/100696862/156997552-3fb9192c-46f0-4084-bdd1-9293efbb8532.png)

**Comment by rpautrat** [comment]
Hi, no this does not look normal. Did you try to visualize the exported ground truth generated by step 3? You can do so using this notebook: https://github.com/cvg/SOLD2/blob/main/notebooks/visualize_exported_dataset.ipynb

**Comment by Zhuyawen7** [comment]
hi,
I tried the "visualize_exported_dataset.ipynb", but I have this problem.There is no ref_image in the data1.

 
![image](https://user-images.githubusercontent.com/100696862/157042126-b00dab73-212f-458b-b0fb-307d9703506f.png)

**Comment by rpautrat** [comment]
Hi, in the config file [sold2/config/wireframe_dataset.yaml](https://github.com/cvg/SOLD2/blob/main/sold2/config/wireframe_dataset.yaml), you can change the 'return_type' to 'paired_desc' and that should solve it.

**Comment by Zhuyawen7** [comment]
> try to visualize the exported ground truth generated
Hi, thanks for solving my last problem!
I tried to visualize the exported ground truth generated  by step 3, it looks good. In step 4, now i am training in 170epoch, but the "Heatmap" recall is only 0.0552. Is this normal please?
![image](https://user-images.githubusercontent.com/100696862/157622470-30162f74-e535-4f19-9d44-5db1786b4c7c.png)

**Comment by rpautrat** [comment]
Hi, yes, the heatmap recall was also very low for us (around 0.06~0.08 at the end). Again, this is more a metric to see the progress of the learning, but it is too low-level to trust it completely.

Regarding your other metrics, I think they can be improved by training longer. On the validation set, we had a junction precision NMS of ~0.3, junction recall NMS of ~0.35, heatmap precision of ~0.62 and heatmap recall ~0.06. These numbers are for the joint model with lines and descriptors. The model trained with lines only had slightly better metrics.

**Comment by Zhuyawen7** [comment]
Thanks for your help, I've done the training, but the results don't seem to make a big difference.
May I ask how to visualize the matching of LSD+LLD and LSD+WLD? Like the figure 10 in your paper.
